WorkbookRepository tests

aunger opened the pull request:
1. Modify WorkbookRepository.kt so its database calls can be mocked more cleanly
2. Unit-test the Workbook data model, its reactive structures, and DB read and reactive write calls
3. Fix bugs uncovered by testing

This PR depends on minor changes introduced in https://github.com/WycliffeAssociates/otter-common/pull/98

aunger committed:
Prepare WorkbookRepository for DB mocking.

aunger committed:
WorkbookRepository unit tests.

aunger committed:
Fix lazy loading. (Thank you, tests!)

aunger committed:
Remove unused Powermock library.

aunger committed:
Add MockitoKotlin2 library.

jsarabia commented on src/main/kotlin/org/wycliffeassociates/otter/jvm/persistence/repositories/WorkbookRepository.kt:
```diff
@@ -40,54 +41,53 @@ class WorkbookRepository(
-            .concatMapEager { constructChapter(it).toObservable() }
-
-        return chapters.cache()
+        val coldObservable = Observable.fromCallable {
```
how about "sortedChapters" as a name here?

jsarabia replied:
most things we're doing with rx are cold observables, so my thought was that the current name doesn't provide a lot of information regarding what the observable is doing

aunger replied:
This little dance where I'm wrapping observable inside observable is in order to keep them from going hot. Some operations like flatmaps seem to make them go hot (revealed by the tests in this pr). So the purpose of this wrapper is to keep everything *cold* and be lazy. I can see it's a weird name though and can rename it.

aunger replied:
I couldn't fix this by switching to map or concatMap from concatMapEager. However, the Observable.defer function does what I needed more cleanly, so I reverted this change and implemented that.

jsarabia commented on src/main/kotlin/org/wycliffeassociates/otter/jvm/persistence/repositories/WorkbookRepository.kt:
```diff
@@ -40,54 +41,53 @@ class WorkbookRepository(
-            }
-
-        return chunks.cache()
+        val coldObservable = Observable.fromCallable {
```
same as above, how about "sortedChunks"?

aunger replied:
Switched to Observable.defer

KJoslyn commented on src/main/kotlin/org/wycliffeassociates/otter/jvm/persistence/repositories/WorkbookRepository.kt:
```diff
@@ -40,54 +41,53 @@ class WorkbookRepository(
-        val contents = contentRepository
-            .getByCollection(chapterCollection)
-            .flattenAsObservable { list -> list.sortedBy { it.sort } }
-            .filter { it.labelKey != "chapter" } // TODO: filter by something better
```
I believe this should be
`.filter { it.labelKey == "verse" }`
As it is now, this function will retrieve title and body contents as well.

aunger replied:
Changed.

KJoslyn committed:
Fix for getResources

aunger committed:
Revert "Fix lazy loading. (Thank you, tests!)"

This reverts commit 299ce676915270afe9778c1cfdc9a013511668ef.
We think there is a better way to solve this.

aunger committed:
Fix lazy loading, et al:

Fix laxy loading with Observable.defer.
Fix Observable.buffer.
Fix PR comment about filtering by verse content type.
Push sorting into DBMS.

aunger committed:
Move WorkbookRepository from otter-jvm to otter-common repository.

aunger committed:
Undo IDEA's auto-formatting. No, thank you, IDEA.

jsarabia commented on src/main/kotlin/org/wycliffeassociates/otter/jvm/persistence/repositories/WorkbookRepository.kt:
```diff
@@ -40,56 +41,49 @@ class WorkbookRepository(
+            db.getChildren(bookCollection)
+                .flattenAsObservable { it }
+                .concatMapEager { constructChapter(it).toObservable() }
+        }.cache()
```
Does defer work with cache?

aunger replied:
Yes, I believe this is verified by the tests chaptersAreLazyLoad and chaptersAreCached.

jsarabia replied:
I was curious because the docs made it sound like it creates a new observable every time it's subscribed, but I wasn't sure that cache stopped that behavior. Observable.fromCallable is also supposed to be lazy and defer execution until subscribe, but made no specific mention of making a new observable every time.

I also didn't see the tests, so I appreciate you mentioning it. They got grouped up into one small "this diff is too large to display by default" thing and I totally missed it. Which definitely made me confused as to the name of the PR ahaha.

aunger commented:
Yes, I believe this is verified by the tests chaptersAreLazyLoad and chaptersAreCached.

aunger committed:
Merge pull request #452 from WycliffeAssociates/rba-move-workbook-to-common

Move WorkbookRepository to common
